### PR TITLE
[FluentTheme] Addition of ColorPaletteResources and automatic system accent resource colors

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

Status: API is ready but doesn't work properly because of another static resources bug.
TBD, going to write a full documentation here, which will be reused in the actual docs page later.

```xaml
<FluentTheme>
  <FluentTheme.Palettes>
    <!-- redefines all colors of the Fluent Light variant -->
    <ColorPaletteResources x:Key="Default"
                           Accent="#FF34854D" AltHigh="#FFFFFFFF" AltLow="#FFFFFFFF"
                           AltMedium="#FFFFFFFF" AltMediumHigh="#FFFFFFFF" AltMediumLow="#FFFFFFFF"
                           BaseHigh="#FF000000" BaseLow="#FFC2DB65" BaseMedium="#FF7D9728" BaseMediumHigh="#FF4F6A00"
                           BaseMediumLow="#FF668114" ChromeAltLow="#FF4F6A00" ChromeBlackHigh="#FF000000"
                           ChromeBlackLow="#FFC2DB65" ChromeBlackMedium="#FF4F6A00" ChromeBlackMediumLow="#FF7D9728"
                           ChromeDisabledHigh="#FFC2DB65" ChromeDisabledLow="#FF7D9728" ChromeGray="#FF668114"
                           ChromeHigh="#FFC2DB65" ChromeLow="#FFE6F3BB" ChromeMedium="#FFDFEEAA"
                           ChromeMediumLow="#FFE6F3BB" ChromeWhite="#FFFFFFFF" ListLow="#FFDFEEAA"
                           ListMedium="#FFC2DB65" />
    <!-- will use default values for other coloors -->
    <ColorPaletteResources x:Key="Dark"
                           Accent="#FF34854D" />
  </FluentTheme.Palettes>
</FluentTheme>
```

Yes, it is compatible with output of [Fluent XAML Theme Editor](https://github.com/microsoft/fluent-xaml-theme-editor), if anybody remembers it ([store link](https://apps.microsoft.com/store/detail/fluent-xaml-theme-editor/9N2XD3Q8X57C?hl=en-us&gl=us)). With some minor edits like xmlns removal. Also in Avalonia ColorPaletteResources can be used only on FluentTheme.Palettes (i.e. globally), avoiding bugs like UWP has, when users attempt to use it in the middle of the visual tree.

## What is the current behavior?

There is no easy way to redefine some base colors of the FluentTheme.
FluentTheme still uses hardcoded blue accent color (we have low level API already and it was used in the ControlCatalog for a while, but this PR moves this code to the FluentTheme package).

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes

Internal structure of FluentTheme was once again changed. But hopefully nobody will need to use it directly anymore (unless some extremely special cases).

## Fixed issues

TBD

### CODE REVIEW

File: src/Avalonia.Themes.Fluent/Accents/SystemAccentColors.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            
+            _systemAccentColor = _platformSettings?.GetColorValues().AccentColor1 ?? s_defaultSystemAccentColor;
+            (_systemAccentColorDark1,_systemAccentColorDark2, _systemAccentColorDark3,
+                    _systemAccentColorLight1, _systemAccentColorLight2, _systemAccentColorLight3) = CalculateAccentShades(_systemAccentColor);
+        }
+    }
+
+    public static (Color d1, Color d2, Color d3, Color l1, Color l2, Color l3) CalculateAccentShades(Color accentColor)
+    {
+        // dark1step = (hslAccent.L - SystemAccentColorDark1.L) * 255
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Accents/SystemAccentColors.cs:134]: If anybody interested, I was able to replicate close to identical algorithm of getting accent color shades.
(unless it's different depending on the color)

File: src/Avalonia.Themes.Fluent/ColorPaletteResources.Properties.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    public Color Accent
+    {
+        get => _accentColor;
+        set => SetAndRaise(AccentProperty, ref _accentColor, value);
+    }
+
+    /// <summary>
+    /// Gets or sets the AltHigh color value.
+    /// </summary>
+    public Color AltHigh { get => GetColor("SystemAltHighColor"); set => SetColor("SystemAltHighColor", value); }
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/ColorPaletteResources.Properties.cs:27]: Only accent color is bindable, as it is used with DynamicResource everywhere.
Other colors are not bindable, and I consider adding some kind of setter validation to throw an exception, if it was setter after resources were read once.

**Comment by maxkatz6** [comment]
And no, I am not considering adding anything like this to the SimpleTheme.
We can potentially have a special dictionary for the highlight colors there (equivalent of accent colors in fluent).
But I am going to leave it to the community, if anybody needs it.

ColorPaletteResources probably is useless there, as simple theme has a very little number of resources to override.

**Comment by Mrxx99** [comment]
If the `ColorPaletteResources` is specific to FluentTheme, wouldn't it make sense to call it `FluentColorPaletteResources`? So it is clear for which Theme it is and other theme libraries could follow a similar naming scheme if they wanted to.

**Comment by rabbitism** [comment]
My theme library (semi avalonia) is a good example of independent theme. How would I use this mechanism? What would happen if someone wants to switch from fluent to semi?

**Comment by grokys** [comment]
> How would I use this mechanism?

I'm not the author but I expect it would be best to copy the code into your theme for the moment. I suspect it may be something we want to expose as a general-purpose public API at some point, but I imagine we'd want to spend some time tweaking it as part of the fluent theme for a while first.

**Comment by rabbitism** [comment]
@grokys sorry I misunderstood this PR. It is reasonable to have this specific to fluent theme class.

File: src/Avalonia.Themes.Fluent/Accents/SystemAccentColors.cs (hunk 1 above)
Comment (robloo) [reply src/Avalonia.Themes.Fluent/Accents/SystemAccentColors.cs:134]: A few years ago some of us tried to figure this out as well. Nobody thought at the time to use HSL but we did use HSV. In that HSV analysis weightings were changed based on the color. The algorithm seems adaptive to perception for each color. Maybe HSL really is the solution though.

Here is some of the old discussion.
https://github.com/CommunityToolkit/WindowsCommunityToolkit/issues/3363#issuecomment-650562951

Whatever is done here for the Fluent color palette really should also be done in the ColorPicker's previewer control. The intention was for that to work the same as what you are doing here.

**Review by grokys — APPROVED** [review]
LGTM  as something that can be built on! We definitely need something like this.

**Pull request merged**